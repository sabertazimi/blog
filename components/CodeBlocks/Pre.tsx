import cx from 'classnames'
import dynamic from 'next/dynamic'
import type { HTMLProps, ReactElement } from 'react'
import BlockCode from './BlockCode'
import styles from './Pre.module.css'
import {
  normalizeCode,
  normalizeLanguage,
  normalizeLanguageName,
  normalizeLines,
} from './utils'
import CopyButton from '@/components/CopyButton'

const LiveCode = dynamic(() => import('./LiveCode'))

interface Props extends HTMLProps<HTMLPreElement> {
  live?: boolean
  noline?: boolean
  nocopy?: boolean
  title?: string
  lines?: string
}

interface CodeProps {
  children?: string
  className?: string
}

function Pre({
  live = false,
  noline = false,
  nocopy = false,
  title = '',
  lines = '',
  children,
  className,
}: Props): JSX.Element {
  const codeElement = children as ReactElement<CodeProps>
  const code = normalizeCode(codeElement?.props?.children)
  const languageClass = codeElement?.props?.className
  const language = normalizeLanguage(languageClass)
  const languageName = normalizeLanguageName(language)
  const highlightLines = normalizeLines(lines)

  return (
    <pre
      className={cx(
        className,
        languageClass,
        styles.pre,
        'dark:shadow-xl dark:shadow-primary',
      )}
      data-language={title || languageName}
    >
      {!nocopy && !live ? <CopyButton code={code} /> : null}
      {live
        ? (
          <LiveCode language={language}>{code}</LiveCode>
          )
        : (
          <BlockCode
            enableLine={!noline}
            lines={highlightLines}
            language={language}
          >
            {code}
          </BlockCode>
          )}
    </pre>
  )
}

export default Pre
