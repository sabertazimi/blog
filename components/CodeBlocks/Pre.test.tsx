import { render } from '@testing-library/react';
import Pre from './Pre';

describe('Pre', () => {
  const languages = [
    'html',
    'xml',
    'css',
    'json',
    'md',
    'markdown',
    'js',
    'javascript',
    'ts',
    'typescript',
    'cs',
    'coffeescript',
    'jsx',
    'tsx',
    'oc',
    'objectivec',
    'vue',
    'go',
    'java',
    'rust',
  ];

  test('should render correctly (snapshot)', () => {
    const { container } = render(<Pre />);

    expect(container).toMatchSnapshot();
  });

  test('should render children correctly (snapshot)', () => {
    const { container } = render(
      <Pre>
        <code>Code</code>
      </Pre>
    );

    expect(container).toMatchSnapshot();
  });

  test.each(languages)(
    'should render language correctly (snapshot)',
    language => {
      const { container } = render(
        <Pre>
          <code className={`language-${language}`}>Code</code>
        </Pre>
      );

      expect(container).toMatchSnapshot();
    }
  );
});
