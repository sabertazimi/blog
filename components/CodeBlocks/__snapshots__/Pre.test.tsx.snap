// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pre should render children correctly (snapshot) 1`] = `
<div>
  <pre
    class="pre"
    data-language="Code"
  >
    <code>
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render correctly (snapshot) 1`] = `
<div>
  <pre
    class="pre"
    data-language="Code"
  />
</div>
`;

exports[`Pre should render language correctly (snapshot) 1`] = `
<div>
  <pre
    class="language-html pre"
    data-language="HTML"
  >
    <code
      class="language-html"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 2`] = `
<div>
  <pre
    class="language-xml pre"
    data-language="XML"
  >
    <code
      class="language-xml"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 3`] = `
<div>
  <pre
    class="language-css pre"
    data-language="CSS"
  >
    <code
      class="language-css"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 4`] = `
<div>
  <pre
    class="language-json pre"
    data-language="JSON"
  >
    <code
      class="language-json"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 5`] = `
<div>
  <pre
    class="language-md pre"
    data-language="Markdown"
  >
    <code
      class="language-md"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 6`] = `
<div>
  <pre
    class="language-markdown pre"
    data-language="Markdown"
  >
    <code
      class="language-markdown"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 7`] = `
<div>
  <pre
    class="language-js pre"
    data-language="JavaScript"
  >
    <code
      class="language-js"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 8`] = `
<div>
  <pre
    class="language-javascript pre"
    data-language="JavaScript"
  >
    <code
      class="language-javascript"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 9`] = `
<div>
  <pre
    class="language-ts pre"
    data-language="TypeScript"
  >
    <code
      class="language-ts"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 10`] = `
<div>
  <pre
    class="language-typescript pre"
    data-language="TypeScript"
  >
    <code
      class="language-typescript"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 11`] = `
<div>
  <pre
    class="language-cs pre"
    data-language="CoffeeScript"
  >
    <code
      class="language-cs"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 12`] = `
<div>
  <pre
    class="language-coffeescript pre"
    data-language="CoffeeScript"
  >
    <code
      class="language-coffeescript"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 13`] = `
<div>
  <pre
    class="language-jsx pre"
    data-language="React"
  >
    <code
      class="language-jsx"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 14`] = `
<div>
  <pre
    class="language-tsx pre"
    data-language="React"
  >
    <code
      class="language-tsx"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 15`] = `
<div>
  <pre
    class="language-oc pre"
    data-language="Objective-C"
  >
    <code
      class="language-oc"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 16`] = `
<div>
  <pre
    class="language-objectivec pre"
    data-language="Objective-C"
  >
    <code
      class="language-objectivec"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 17`] = `
<div>
  <pre
    class="language-vue pre"
    data-language="Vue"
  >
    <code
      class="language-vue"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 18`] = `
<div>
  <pre
    class="language-go pre"
    data-language="Go"
  >
    <code
      class="language-go"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 19`] = `
<div>
  <pre
    class="language-java pre"
    data-language="Java"
  >
    <code
      class="language-java"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Pre should render language correctly (snapshot) 20`] = `
<div>
  <pre
    class="language-rust pre"
    data-language="Rust"
  >
    <code
      class="language-rust"
    >
      Code
    </code>
  </pre>
</div>
`;
